File: src/main/java/me/ag2s/epublib/util/ResourceUtil.java
package me.ag2s.epublib.util;

import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.Reader;
import java.io.UnsupportedEncodingException;
import java.util.zip.ZipEntry;
import java.util.zip.ZipInputStream;
import javax.xml.parsers.DocumentBuilder;
import me.ag2s.epublib.domain.MediaType;
import me.ag2s.epublib.domain.MediaTypes;
import me.ag2s.epublib.domain.Resource;
import me.ag2s.epublib.epub.EpubProcessorSupport;
import org.w3c.dom.Document;
import org.xml.sax.InputSource;
import org.xml.sax.SAXException;

public class ResourceUtil {
   public static Resource createChapterResource(String title, String txt, String model, String href) {
      if (title.contains("\n")) {
         title = "<span class=\"chapter-sequence-number\">" + title.replaceFirst("\\s*\\n\\s*", "</span><br />");
      } else {
         title = title.replaceFirst("\\s+", "</span><br />");
         if (title.contains("</span>")) {
            title = "<span class=\"chapter-sequence-number\">" + title;
         }
      }

      String html = model.replace("{title}", title).replace("{content}", StringUtil.formatHtml(txt));
      return new Resource(html.getBytes(), href);
   }

   public static Resource createPublicResource(String name, String author, String intro, String kind, String wordCount, String model, String href) {
      String html = model.replace("{name}", name).replace("{author}", author).replace("{kind}", kind == null ? "" : kind).replace("{wordCount}", wordCount == null ? "" : wordCount).replace("{intro}", StringUtil.formatHtml(intro == null ? "" : intro));
      return new Resource(html.getBytes(), href);
   }

   public static Resource createResource(File file) throws IOException {
      if (file == null) {
         return null;
      } else {
         MediaType mediaType = MediaTypes.determineMediaType(file.getName());
         byte[] data = IOUtil.toByteArray(new FileInputStream(file));
         return new Resource(data, mediaType);
      }
   }

   public static Resource createResource(String title, String href) {
      String content = "<html><head><title>" + title + "</title></head><body><h1>" + title + "</h1></body></html>";
      return new Resource((String)null, content.getBytes(), href, MediaTypes.XHTML, "UTF-8");
   }

   public static Resource createResource(ZipEntry zipEntry, ZipInputStream zipInputStream) throws IOException {
      return new Resource(zipInputStream, zipEntry.getName());
   }

   public static Resource createResource(ZipEntry zipEntry, InputStream zipInputStream) throws IOException {
      return new Resource(zipInputStream, zipEntry.getName());
   }

   public static byte[] recode(String inputEncoding, String outputEncoding, byte[] input) throws UnsupportedEncodingException {
      return (new String(input, inputEncoding)).getBytes(outputEncoding);
   }

   public static InputSource getInputSource(Resource resource) throws IOException {
      if (resource == null) {
         return null;
      } else {
         Reader reader = resource.getReader();
         return reader == null ? null : new InputSource(reader);
      }
   }

   public static Document getAsDocument(Resource resource) throws SAXException, IOException {
      return getAsDocument(resource, EpubProcessorSupport.createDocumentBuilder());
   }

   public static Document getAsDocument(Resource resource, DocumentBuilder documentBuilder) throws UnsupportedEncodingException, SAXException, IOException {
      InputSource inputSource = getInputSource(resource);
      return inputSource == null ? null : documentBuilder.parse(inputSource);
   }
}

File: src/main/kotlin/me/ag2s/epublib/util/ResourceUtil.kt
package me.ag2s.epublib.util
import java.io.*
import java.util.zip.*
import javax.xml.parsers.DocumentBuilder
import me.ag2s.epublib.domain.*
import me.ag2s.epublib.epub.EpubProcessorSupport
import org.w3c.dom.Document
import org.xml.sax.InputSource
import java.nio.charset.Charset

object ResourceUtil {
    @JvmStatic fun createChapterResource(title:String, txt:String, model:String, href:String):Resource { val formatted=if(title.contains("\n")) "<span class=\"chapter-sequence-number\">"+title.replaceFirst(Regex("\\s*\\n\\s*"),"</span><br />") else title.replaceFirst(Regex("\\s+"),"</span><br />").let { if(it.contains("</span>")) "<span class=\"chapter-sequence-number\">$it" else it }; return Resource(model.replace("{title}",formatted).replace("{content}",StringUtil.formatHtml(txt)).toByteArray(),href) }
    @JvmStatic fun createPublicResource(name:String,author:String,intro:String?,kind:String?,wordCount:String?,model:String,href:String):Resource = Resource(model.replace("{name}",name).replace("{author}",author).replace("{kind}",kind?:"").replace("{wordCount}",wordCount?:"").replace("{intro}",StringUtil.formatHtml(intro?:"")).toByteArray(),href)
    @JvmStatic fun createResource(file:File?):Resource? = file?.let { Resource(IOUtil.toByteArray(FileInputStream(it)),MediaTypes.determineMediaType(it.name)) }
    @JvmStatic fun createResource(title:String,href:String):Resource = Resource(null,"<html><head><title>$title</title></head><body><h1>$title</h1></body></html>".toByteArray(),href,MediaTypes.XHTML,"UTF-8")
    @JvmStatic fun createResource(entry:ZipEntry,input:ZipInputStream):Resource = Resource(input,entry.name)
    @JvmStatic fun createResource(entry:ZipEntry,input:InputStream):Resource = Resource(input,entry.name)
    @JvmStatic fun recode(inputEncoding:String,outputEncoding:String,input:ByteArray):ByteArray = String(input,Charset.forName(inputEncoding)).toByteArray(Charset.forName(outputEncoding))
    @JvmStatic fun getInputSource(resource:Resource?):InputSource? = resource?.getReader()?.let(::InputSource)
    @JvmStatic fun getAsDocument(resource:Resource?):Document? = getAsDocument(resource,EpubProcessorSupport.createDocumentBuilder()!!)
    @JvmStatic fun getAsDocument(resource:Resource?,builder:DocumentBuilder):Document? = getInputSource(resource)?.let(builder::parse)
}
